### DISCUSSION TITLE
Using this crate in no_std and embedded environments

### DISCUSSION THREAD

**Discussion by itamarst** [anweiss/cddl#119] [Feedback]
I am wrapping the library with Python, using PyO3. PyO3's macros for wrapping classes aren't happy with having a lifetime parameter, so having a `CDDL` attribute is difficult (I don't want to parse the schema from scratch each time I validate a document).

More broadly, having a lifetime identifier on the `CDDL` struct just makes using it annoying since it percolates to everything.

I am very much not an expert, but I suspect this could be fixed simply by having the parser own the lexer and the original schema string?

**Comment by anweiss** [comment]
This is due to the fact that one of my original goals for the project was to implement this in as close to a zero-copy way as possible given CBOR's (and presumably CDDL's) use in constrained computing environments. I assume you're compiling to the WASM target? If so, I can just created owned versions of these structs which are used when targeting WASM. Would that work?

**Comment by anweiss** [comment]
Hey @itamarst, I need to think about this a bit more. Creating owned versions of everything will add quite a bit of duplication and extra work, which I don't feel is worth it just to work around PyO3's limitations for using lifetimes.

**Comment by itamarst** [comment]
1. Given #105, mostly I just care about `CDDL` not having a livetime.
2. I can't really think of any use case for non-owning `CDDL`, so I would just design it to always own the schema bytes, and then you don't have two variants, you just have one. However, I am not really familiar with `no_std` world much, so maybe it makes sense there?
3. If you really want to support both, maybe `Cow` would do the trick?

**Comment by itamarst** [comment]
Oh, `Cow` has a lifetime too. Bleh.

**Comment by anweiss** [comment]
Yea, gonna move this to a Discussion and gauge interest for now. It's unclear, even to me, what the embedded/`no_std` folks actually require from this crate.

**Comment by itamarst** [comment]
What if `CDDL` was:

```rust
struct CDDL<B: AsRef<[u8]>> {...}
```

I think this might support all the options, looking at https://doc.rust-lang.org/std/convert/trait.AsRef.html#implementors?

**Reply by itamarst** [reply]
Hm, no, would still need a lifetime in order to store the underlying structs...

**Comment by itamarst** [comment]
OK, I figured out the solution (without requiring changes on your side): https://docs.rs/ouroboros/latest/ouroboros/attr.self_referencing.html allowed me to store both the `String` I get from Python and the resulting `CDDL` on the same object.

**Reply by anweiss** [reply]
Good find @itamarst! Will keep this discussion open nonetheless to facilitate further dialogue about CDDL in `no_std` environments.
